iree/compiler/Dialect/Flow/Transforms/MatmulToMMT4dConversion.cpp

```diff
@@ -0,0 +1,203 @@
+
+  SmallVector<StringRef> loopAttributeTypes(nloops, "parallel");
+
+  SmallVector<AffineMap> indexingMaps = {
+      inversePermutation(
+          AffineMap::get(nloops, 0, exprs, rewriter.getContext())),
+      AffineMap::getMultiDimIdentityMap(nloops, rewriter.getContext())};
+
+  auto transposedOp = rewriter.create<linalg::GenericOp>(
+      loc, outputTensor.getType(),
+      /*inputs=*/operand, /*outputs=*/outputTensor, indexingMaps,
+      loopAttributeTypes,
+      [&](OpBuilder &nestedBuilder, Location nestedLoc, ValueRange args) {
+        nestedBuilder.create<linalg::YieldOp>(nestedLoc, args[0]);
+      });
+
+  return transposedOp.getResult(0);
+};
+
+// Collapses a 4d tensor operand to 2d given its target shape.
+static Value collapseTo2D(mlir::Location loc, PatternRewriter &rewriter,
+                          Value operand, ArrayRef<int64_t> targetShape) {
+  auto operandType = operand.getType().cast<RankedTensorType>();
+  auto targetType =
+      RankedTensorType::get(targetShape, operandType.getElementType());
+  SmallVector<ReassociationIndices> collapseIndices = {{0, 1}, {2, 3}};
+  Value reshapedOperand = rewriter.create<linalg::TensorCollapseShapeOp>(
+      loc, targetType, operand, collapseIndices);
+  return reshapedOperand;
+}
+
+// Converts linalg.matmul -> linalg.mmt4d where M0, N0, K0 are compile time
+// constants.
+class LinalgMatmulOpToLinalgMMT4dOpPattern
+    : public OpRewritePattern<linalg::MatmulOp> {
+ public:
+  LinalgMatmulOpToLinalgMMT4dOpPattern(MLIRContext *context, int M0, int N0,
+                                       int K0, PatternBenefit benefit = 1)
+      : OpRewritePattern<linalg::MatmulOp>(context, benefit),
+        M0Size(M0),
+        N0Size(N0),
+        K0Size(K0) {}
+
+  LogicalResult matchAndRewrite(linalg::MatmulOp matmulOp,
+                                PatternRewriter &rewriter) const override {
+    auto loc = matmulOp.getLoc();
+
+    Value lhs = matmulOp.getInputOperand(0)->get();
+    Value rhs = matmulOp.getInputOperand(1)->get();
+    Value dst = matmulOp.getOutputOperand(0)->get();
+
+    RankedTensorType lhsType = lhs.getType().dyn_cast<RankedTensorType>();
+    RankedTensorType rhsType = rhs.getType().dyn_cast<RankedTensorType>();
+    RankedTensorType dstType = dst.getType().dyn_cast<RankedTensorType>();
+
+    if (!lhsType || !rhsType || !lhsType.hasStaticShape() ||
+        !rhsType.hasStaticShape()) {
+      return failure();
+    }
+
+    // This is for float only matmul for now. Integer data type might require
+    // r.h.s layout change.
+    if (!lhsType.getElementType().isF32() || !rhsType.getElementType().isF32())
+      return failure();
+
+    int m = lhsType.getShape()[0];
+    int n = rhsType.getShape()[1];
+    int k = rhsType.getShape()[0];
+
+    if (m % M0Size != 0 || n % N0Size != 0 || k % K0Size != 0) return failure();
+
+    int m1 = m / M0Size;
+    int n1 = n / N0Size;
+    int k1 = k / K0Size;
+
+    bool isInitTensorDst = false;
```

Comment:
(I had this comment previously that probably got lost in the ether due to github issues).

I am wondering why the DAG needs to be matched here. The transformation itself should not care where the `outs` tensor is coming from. If this is a "canonicalization" would be better to split it up. If it makes sense in general, this canonicalization could live upstream, or here is fine to as a separate pattern. You should be able to just use `transposeOperand` on the `outs` argument.

Reply:
Ultimately this should be just `transposeOperand(outs` and we will need to added as a canonicalization `([init_tensor -> fill -> transpose] -> [init_tensor -> fill])` upstream, I will add a TODO for that.

Reply:
This is basically `fill -> transpose` to `fill`. I dont see a reason to delay this. We can land that upstream (let me know when you have it, I will review it ASAP). That is a good pattern to have anyway. (Its easy to just forget about TODOs. We have a lot of TODOs)

Reply:
This pattern is temporarily here any and will be upstreamed eventually. In this case this won't be a forgotten TODO :)

Reply:
Ok, can we split this into a separate pattern then. You can just run `applyPatternsAndFoldGreedily` twice, once that runs the matmul -> mmt4d conversion and then the patterns to clean the things up. We can then selectively move those as canonicalizations upstream and drop them here.